src/test/java/teammates/test/driver/EmailAccount.java

```diff
@@ -26,73 +34,204 @@ private EmailAccount() {
+            return null;
+        }
+
+        for (Message messageStub : messageStubs) {
+            final Message message = service.users().messages().get(username, messageStub.getId()).setFormat("raw")
+                    .execute();
+
+            final MimeMessage email = convertFromMessageToMimeMessage(message);
+
+            if (isStudentCourseJoinRegistrationEmail(email, courseName, courseId)) {
+                final String body = getTextFromEmail(email);
+
+                markMessageAsRead(service, username, messageStub);
+
                 return getKey(body);
             }
         }
 
         return null;
     }
 
-    private static boolean isRegistrationEmail(Message message, String courseName, String courseId)
+    private static void markMessageAsRead(Gmail service, String username, Message messageStub) throws IOException {
+        final ModifyMessageRequest modifyMessageRequest = new ModifyMessageRequest()
+                .setRemoveLabelIds(Collections.singletonList("UNREAD"));
+        service.users().messages().modify(username, messageStub.getId(), modifyMessageRequest).execute();
+    }
+
+    private static boolean isEmpty(List<Message> messageStubs) {
+        return messageStubs == null;
+    }
+
+    private static MimeMessage convertFromMessageToMimeMessage(Message message) throws MessagingException {
+        byte[] emailBytes = BaseEncoding.base64Url().decode(message.getRaw());
+
+        // While we are not actually sending or receiving an email, a session is required so there will be strict parsing
+        // of address headers when we create a MimeMessage. We are also passing in empty properties where we are expected to
+        // supply some values because we are not actually sending or receiving any email.
+        Session session = Session.getInstance(new Properties());
+
+        return new MimeMessage(session, new ByteArrayInputStream(emailBytes));
+    }
+
+    private static boolean isStudentCourseJoinRegistrationEmail(MimeMessage message, String courseName, String courseId)
             throws MessagingException {
         String subject = message.getSubject();
         return subject != null && subject.equals(String.format(EmailType.STUDENT_COURSE_JOIN.getSubject(),
-                                                               courseName, courseId));
+                courseName, courseId));
     }
 
-    private static String getKey(String body) {
-        String key = body.substring(
-                body.indexOf("key=") + "key=".length(),
-                body.indexOf("studentemail=") - 1); //*If prompted to log in
-        return key.trim();
+    /**
+     * Gets the email message body as text.
+     */
+    private static String getTextFromEmail(MimeMessage email) throws MessagingException, IOException {
+        if (email.isMimeType("text/*")) {
+            return (String) email.getContent();
+        } else {
+            return getTextFromPart(email);
```

Comment:
It isn't but it's an alternate flow? If I leave that out it would look like I'm handling special cases first.